### PR TITLE
OCPCLOUD-3215: Add support for dual networking stack services

### PR DESCRIPTION
**Pull request by nrb**
Carry the upstream change https://github.com/kubernetes/cloud-provider-aws/pull/1313 in our fork.

### CODE REVIEW

**Comment by shellyyang1989** [comment]
@huali9 for awareness of the downstream change for the aws dual stack feature

File: pkg/providers/v1/aws_loadbalancer_test.go [hunk 1]
```diff
@@ -1921,3 +1945,356 @@ func TestCloud_reconcileTargetGroupsAttributes(t *testing.T) {
...
+					Id:   aws.String("i-abcdefg3"),
+					Port: aws.Int32(8080),
+				},
+				{
+					Id:   aws.String("i-abcdefg4"),
+					Port: aws.Int32(8080),
+				},
+			},
+			description: "Function should succeed when replacing an instance for a LB at max capacity",
+		},
```
[the remark this answers is not in the document]
Comment (nrb) [reply pkg/providers/v1/aws_loadbalancer_test.go:2025]: Acknowledged, but not going to address that with this PR.

**Comment by yunjiang29** [comment]
Install succeeded in:
[e2e-aws-ovn-dualstack-ipv6-primary-techpreview](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/multi-pr-openshift-cluster-ingress-operator-1376-openshift-cloud-provider-aws-135-openshift-installer-10380-openshift-cluster-kube-apiserver-operator-2079-e2e-aws-ovn-dualstack-ipv6-primary-techpreview/2036803341543018496) and [periodic-ci-openshift-openshift-tests-private-release-4.22-amd64-nightly-aws-ipi-byo-subnets-ipv6-tp-f7](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/76791/rehearse-76791-periodic-ci-openshift-openshift-tests-private-release-4.22-amd64-nightly-aws-ipi-byo-subnets-ipv6-tp-f7/2036751120465924096), the payload includes: 
* https://github.com/openshift/cloud-provider-aws/pull/135
* https://github.com/openshift/cluster-kube-apiserver-operator/pull/2079
* https://github.com/openshift/cluster-ingress-operator/pull/1376


cc @huali9 for final review and verification.

**Comment by nrb** [comment]
/testwith openshift/installer/main/e2e-aws-ovn-dualstack-ipv6-primary-techpreview https://github.com/openshift/cluster-ingress-operator/pull/1376

/testwith openshift/installer/main/e2e-aws-ovn-dualstack-ipv4-primary-techpreview https://github.com/openshift/cluster-ingress-operator/pull/1376

Running the dual stack scenario jobs with the ingress operator, as those were failing prior to the latest commit.

**Comment by huali9** [comment]
> cc @huali9 for final review and verification.

Thank you @yunjiang29 I'm working on this and try to complete the testing today.

**Comment by shellyyang1989** [comment]
@nrb NE team is testing it along with all the installer and ingress changes. So we might not need to run it again here. https://github.com/openshift/origin/pull/30904#issuecomment-4131270132

File: pkg/providers/v1/aws_validations_test.go [hunk 2]
```diff
@@ -564,3 +564,94 @@ func TestValidateServiceAnnotations(t *testing.T) {
...
+			expectedError:  "",
+		},
+		{
+			name:           "RequireDualStack with two entries works",
+			ipFamilyPolicy: v1.IPFamilyPolicyPreferDualStack,
+			ipFamilies:     []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			expectedError:  "",
+		},
+		{
+			name:           "RequireDualStack with two entries works",
```
Comment (sadasu) [inline pkg/providers/v1/aws_validations_test.go:624]: Copy paste issue, I think. This and the previous test are identical.

File: pkg/providers/v1/aws_validations_test.go [hunk 3]
```diff
@@ -564,3 +564,94 @@ func TestValidateServiceAnnotations(t *testing.T) {
...
+			expectedError:  "",
+		},
+		{
+			name:           "PreferDualStack with two entries works",
+			ipFamilyPolicy: v1.IPFamilyPolicyPreferDualStack,
+			ipFamilies:     []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			expectedError:  "",
+		},
+		{
+			name:           "PreferDualStack with two entries works",
```
Comment (sadasu) [inline pkg/providers/v1/aws_validations_test.go:612]: Copy paste issue, I think. This and the previous test are identical.

**Comment by huali9** [comment]
I tested creating `RequireDualStack`, `PreferDualStack` and `SingleStack` service on `DualStackIPv4Primary` cluster, no issue. Details are recorded in the test case https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-88428
`DualStackIPv6Primary` cluster installation failed, just found seems another PR https://github.com/openshift/installer/pull/10380 is needed, I'm rebuilding the image and try again.

**Comment by huali9** [comment]
We cannot build image using merged PRs, so I can only build image with openshift/cluster-ingress-operator#1376 and openshift/cloud-provider-aws#135. But the newly built image still doesn't work.

"job [build 4.22,openshift/cluster-ingress-operator#1376,openshift/cloud-provider-aws#135](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/2037049836636737536) succeeded" 
base on 4.22.0-0.nightly-2026-03-25-221249 

"job <https://prow.ci.openshift.org/view/gs/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/2037090450896064512 | build [4.22.0-0.ci](http://4.22.0-0.ci/),openshift/cluster-ingress-operator#1376,openshift/cloud-provider-aws#135> succeeded" 
base on 4.22.0-0.ci-2026-03-25-085725

Neither of these payloads fully contains https://github.com/openshift/installer/pull/10380 and https://github.com/openshift/cluster-kube-apiserver-operator/pull/2079, so seems we have to wait.

**Comment by nrb** [comment]
@huali9 Could we build with a nightly and this PR? Looks like the other 2 merged.

**Comment by huali9** [comment]
>"job [build 4.22,openshift/cluster-ingress-operator#1376,openshift/cloud-provider-aws#135](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/2037049836636737536) succeeded"
base on 4.22.0-0.nightly-2026-03-25-221249

@nrb This one is built with a nightly, but the DualStackIPv6Primary cluster installation failed. Seems 4.22.0-0.nightly-2026-03-25-221249 doesn't contain https://github.com/openshift/installer/pull/10380

**Comment by patrickdillon** [comment]
> > "job [build 4.22,openshift/cluster-ingress-operator#1376,openshift/cloud-provider-aws#135](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/2037049836636737536) succeeded"
> > base on 4.22.0-0.nightly-2026-03-25-221249
> 
> @nrb This one is built with a nightly, but the DualStackIPv6Primary cluster installation failed. Seems 4.22.0-0.nightly-2026-03-25-221249 doesn't contain [openshift/installer#10380](https://github.com/openshift/installer/pull/10380)

The installer PR should not be necessary (just validation), but that nightly is missing the necessary https://github.com/openshift/cluster-kube-apiserver-operator/pull/2079 which is in [4.22.0-0.nightly-2026-03-26-093013](https://amd64.ocp.releases.ci.openshift.org/releasestream/4.22.0-0.nightly/release/4.22.0-0.nightly-2026-03-26-093013)

**Comment by nrb** [comment]
Clusters configured with `DualStackIPv6Primary` will require a change in provisioning services. Specifically, the `spec.ipFamilyPolicy` field needs to be set to `PreferDualStack`.

This is because the Kubernetes default behavior when `spec.ipFamilyPolicy` is undefined is to default it to `SingleStack`, which was the behavior Services had prior to dual stack support. Unfortunately, AWS does not support IPv6 single stack load balancers - they must be dual stack.

@alebedev87 double checked with someone on the networking QE team and noted that the `SingleStack` defaulting happened on vSphere, too, confirming it's not specific to this change.

Changing this default will likely require significant effort upstream and is unlikely to be accepted. Thus, we should make sure we document the behavior when this ships in OCP 4.22 as well as in the upstream PR.


https://github.com/openshift/origin/pull/30938 implements this change for our own origin tests, so this change won't give passing CI signal til that is merged.

**Comment by sadasu** [comment]
/testwith openshift/installer/main/e2e-aws-ovn-dualstack-ipv6-primary-techpreview https://github.com/openshift/installer/pull/10380 https://github.com/openshift/cluster-kube-apiserver-operator/pull/2079 https://github.com/openshift/cluster-ingress-operator/pull/1376 https://github.com/openshift/origin/pull/30934 https://github.com/openshift/origin/pull/30938

File: pkg/providers/v1/aws_validations_test.go (hunk 3 above)
Comment (nrb) [reply pkg/providers/v1/aws_validations_test.go:612]: fixed, will push up after e2es run to prevent confusion

**Comment by sadasu** [comment]
/testwith openshift/installer/main/e2e-aws-ovn-dualstack-ipv4-primary-techpreview https://github.com/openshift/installer/pull/10380 https://github.com/openshift/cluster-kube-apiserver-operator/pull/2079 https://github.com/openshift/cluster-ingress-operator/pull/1376 https://github.com/openshift/origin/pull/30934 https://github.com/openshift/origin/pull/30938

File: pkg/providers/v1/aws.go [hunk 4]
```diff
@@ -2348,9 +2404,18 @@ func (c *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, apiS
...
+	// If no source ranges specified, add defaults based on service IP families
+	// This should be populated by GetLoadBalancerSourceRanges most of the time.
+	if len(sourceCIDRs) == 0 {
+		sourceCIDRs = append(sourceCIDRs, "0.0.0.0/0")
+	}
+
+	// Add IPv6 default range if service supports IPv6.
+	if serviceRequestsIPv6(apiService) && !contains(sourceCIDRs, "::/0") {
+		sourceCIDRs = append(sourceCIDRs, "::/0")
 	}
```
Comment (tthvo) [inline pkg/providers/v1/aws.go:2419]: I guess this block will always force `::/0` even if the user defines a restricted set of allowed CIDRs via:
- `service.spec.loadBalancerSourceRanges`
- annotation `service.beta.kubernetes.io/load-balancer-source-ranges`

Reference: https://github.com/kubernetes/cloud-provider/blob/2e08bbe9b7f1623207c046a892270baf27599b84/service/helpers/helper.go#L61-L85

I think `!contains(sourceCIDRs, "::/0")` should be replaced with `there is no defined allowed IPv6 CIDR`, right?

File: pkg/providers/v1/aws.go [hunk 5]
```diff
@@ -2483,6 +2549,10 @@ func (c *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, apiS
 		}
 		if len(sourceRangeCidrs) == 0 {
 			sourceRangeCidrs = append(sourceRangeCidrs, "0.0.0.0/0")
+			// Add IPv6 default range if service supports IPv6
+		}
+		if serviceRequestsIPv6(apiService) && !contains(sourceRangeCidrs, "::/0") {
+			sourceRangeCidrs = append(sourceRangeCidrs, "::/0")
```
Comment (tthvo) [inline pkg/providers/v1/aws.go:2555]: Same question as above ☝️

File: pkg/providers/v1/aws.go (hunk 4 above)
Comment (tthvo) [reply pkg/providers/v1/aws.go:2419]: Maybe, we can:

```go
ec2SourceRanges, ec2Ipv6SourceRanges := separateIPv4AndIPv6CIDRs(sourceCIDRs)
```

And check if `len(ec2Ipv6SourceRanges) == 0` here, then add `::/0`.

File: pkg/providers/v1/aws_loadbalancer.go [hunk 6]
```diff
@@ -1044,23 +1130,49 @@ func (c *Cloud) updateInstanceSecurityGroupsForNLB(ctx context.Context, lbName s
 	return nil
 }
 
+// isIPv6CIDR returns true if the given CIDR is an IPv6 CIDR.
+// It uses netip.ParsePrefix to properly parse and validate the CIDR notation.
+func isIPv6CIDR(cidr string) bool {
+	prefix, err := netip.ParsePrefix(cidr)
+	if err != nil {
+		// If parsing fails, fall back to simple string check for backward compatibility
+		// This shouldn't happen with valid AWS CIDR blocks, but we handle it gracefully
```
Comment (tthvo) [inline pkg/providers/v1/aws_loadbalancer.go:1139]: nit: we can simplify this by using `"k8s.io/utils/net"`, which exposes `IsIPv6CIDRString` 😁

Comment (tthvo) [reply pkg/providers/v1/aws_loadbalancer.go:1139]: Example: https://github.com/openshift/installer/blob/423f3022c3b9aac824a29db845302f4a1c8b8736/pkg/types/validation/installconfig.go#L1712

File: pkg/providers/v1/aws_validations.go [hunk 7]
```diff
@@ -150,3 +151,87 @@ func validateServiceAnnotationTargetGroupAttributes(v *awsValidationInput) error
...
+	if policy == nil {
+		// Implicit SingleStack: acceptable only if no IPv6 family is present.
+		return !serviceRequestsIPv6(service)
+	}
+
+	switch *policy {
+	case v1.IPFamilyPolicySingleStack:
+		return !serviceRequestsIPv6(service)
+	case v1.IPFamilyPolicyPreferDualStack:
+		return true
```
Comment (tthvo) [inline pkg/providers/v1/aws_validations.go:179]: > SingleStack: Single-stack service. The control plane allocates a cluster IP for the Service, using the first configured service cluster IP range.
>
> PreferDualStack: Allocates both IPv4 and IPv6 cluster IPs for the Service when dual-stack is enabled. If dual-stack is not enabled or supported, it falls back to single-stack behavior.

Reading from docs, my understanding is that if dual-stack is not configured, `PreferDualstack` will fall back to the primary IP family (i.e. the first configured service cluster IP range). That is:

```
ipFamilyPolicy: PreDualStack
ipFamilies: ["IPv6", "IPv4"] # Will fall back to IPv6
```


So, we should also check if the primary IP (i.e. `IPFamilies[0]`) is IPv4, right? I guess it does not matter much because we never support IPV6-only AWS, but maybe in the "future".

**Review by tthvo — COMMENTED** [review]
Sorry my comments came a bit late. I just have a few (mostly for edge cases) questions. This looks good overall to me 👍

**Comment by huali9** [comment]
>job <https://prow.ci.openshift.org/view/gs/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/2037305972635996160 | build [4.22.0-0.ci](http://4.22.0-0.ci/),openshift/cloud-provider-aws#135,openshift/origin#30938> succeeded

DualStackIPv6Primary cluster install successfully with this build and I'm still testing this, will update the result later.

**Comment by huali9** [comment]
I tested creating services with the following configurations across both `DualStackIPv4Primary` and `DualStackIPv6Primary` clusters:
                                                                                                                                    
  - ipFamilyPolicy: RequireDualStack, PreferDualStack, and SingleStack (both IPv4 and IPv6)                                         
  - ipFamilies ordering: ["IPv4","IPv6"] vs ["IPv6","IPv4"]                                                                         
                                                                                                                                    
 All test details are documented in the test case:
  https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-88428 

 Test Results                                                                                                                      
                                                                                                                                    
  On `DualStackIPv4Primary` Cluster:                                                                                                  
                  
  -  ipFamilies: ["IPv4","IPv6"] - Works successfully                                                                             
  -  ipFamilies: ["IPv6","IPv4"] - **Fails with the following CCM error:**
  Error syncing load balancer: failed to ensure load balancer: error trying to register targets in target group:                    
  "operation error Elastic Load Balancing v2: RegisterTargets, https response error StatusCode: 400,                                
  RequestID: e5f8406b-9edb-40f0-a015-25cf74579f32, InvalidTarget: The following targets cannot be used                              
  with this IPv6 target group because of missing primary IPv6 addresses: i-08043fdfb5bf3cfb7,                                       
  i-0fdf34fe17c5bcea2, i-01aeef3411b4976b9, i-01b14c6d42cb95566, i-046f553fbd2f3b4e6, i-0696afe3e41709a8f"                          
                                                                                                                                    
  On `DualStackIPv6Primary` Cluster:                                                                                                  
                                                                                                                                    
  -  ipFamilies: ["IPv4","IPv6"] - Works successfully, LoadBalancer hostname is directly accessible                               
  -  ipFamilies: ["IPv6","IPv4"] - Service is created successfully, **but direct access to the LoadBalancer hostname fails. Access only works through a Route.**                                                            
                  
  Questions                                                                                                                                                                                              
  
 Are these two behavioral differences expected?
                
  1. On `DualStackIPv4Primary` cluster: Should ipFamilies: ["IPv6","IPv4"] fail with the "missing primary IPv6 addresses" error, or should it succeed? I'm uncertain whether this failure is expected because on the DualStackIPv6Primary cluster, both ["IPv4","IPv6"] and ["IPv6","IPv4"] orderings successfully create the service without errors.
  
  2. On `DualStackIPv6Primary` cluster: Should ipFamilies: ["IPv6","IPv4"] require a Route for external access instead of allowing direct LoadBalancer hostname access? This behavior differs from all other tested scenarios where the LoadBalancer hostname is directly accessible.

I'm uncertain whether I should add the verified label based on these results. Please advise on the expected behavior. Thanks!

**Comment by huali9** [comment]
Additional Finding: I thought of another test scenario - modifying an existing service's ipFamilyPolicy from SingleStack to dual-stack, or vice versa. When attempting either modification, the operation fails with an IAM permission error:                 
  elasticloadbalancing:SetIpAddressType is not authorized.                                                         
                                                                                                                                    
  Example - modifying from dual-stack to SingleStack:
 ```
 $ oc edit svc httpd-require-dualstack-ipv4first                                                                                   
  service/httpd-require-dualstack-ipv4first edited
```
                                                                                                                                    
  CCM logs show:     
                                                                                                               
```
I0327 07:04:35.462920       1 aws_loadbalancer.go:273] Updating load balancer a3bf33c22202c4ac2b80b65711b31fd0 IpAddressType from dualstack to ipv4 for default/httpd-require-dualstack-ipv4first
E0327 07:04:35.479465       1 controller.go:302] "Unhandled Error" err="error processing service default/httpd-require-dualstack-ipv4first (retrying with exponential backoff): failed to ensure load balancer: error updating load balancer IpAddressType: \"operation error Elastic Load Balancing v2: SetIpAddressType, https response error StatusCode: 403, RequestID: df28a454-ab70-4b54-b411-bb520e250f1f, api error AccessDenied: User: arn:aws:sts::301721915996:assumed-role/ci-op-txlgv59x-62174-pkh7p-master-role/i-08043fdfb5bf3cfb7 is not authorized to perform: elasticloadbalancing:SetIpAddressType on resource: arn:aws:elasticloadbalancing:ap-northeast-1:301721915996:loadbalancer/net/a3bf33c22202c4ac2b80b65711b31fd0/1bc9d7f024f32fda because no identity-based policy allows the elasticloadbalancing:SetIpAddressType action\"" logger="UnhandledError"
I0327 07:04:35.479514       1 event.go:389] "Event occurred" object="default/httpd-require-dualstack-ipv4first" fieldPath="" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: error updating load balancer IpAddressType: \"operation error Elastic Load Balancing v2: SetIpAddressType, https response error StatusCode: 403, RequestID: df28a454-ab70-4b54-b411-bb520e250f1f, api error AccessDenied: User: arn:aws:sts::301721915996:assumed-role/ci-op-txlgv59x-62174-pkh7p-master-role/i-08043fdfb5bf3cfb7 is not authorized to perform: elasticloadbalancing:SetIpAddressType on resource: arn:aws:elasticloadbalancing:ap-northeast-1:301721915996:loadbalancer/net/a3bf33c22202c4ac2b80b65711b31fd0/1bc9d7f024f32fda because no identity-based policy allows the elasticloadbalancing:SetIpAddressType action\""
```
                   
                                                                                             
  I'm uncertain whether this is expected behavior or a bug. Should the cluster's IAM role include the elasticloadbalancing:SetIpAddressType permission by default for dual-stack clusters, or is modifying the ipFamilyPolicy of existing services not a supported operation?

**Comment by huali9** [comment]
@tthvo [answered all the questions](https://redhat-internal.slack.com/archives/C06M0G7MT1D/p1774598315921059?thread_ts=1774583127.929819&cid=C06M0G7MT1D) and it makes sense to me. no questions from my side now.

/verified by @huali9

**Comment by damdo** [comment]
@knobunc If possible I'd like to wait for confirmation from upstream this is the right direction (at least getting verbal approval) before merging this carry, for sustainability reasons. @nrb is reaching out to them to confirm

**Comment by knobunc** [comment]
> @knobunc If possible I'd like to wait for confirmation from upstream this is the right direction (at least getting verbal approval) before merging this carry, for sustainability reasons. @nrb is reaching out to them to confirm

No problem.  From our Slack conversation it sounds like you are talking to them at Kubecon, so hopefully this gets resolved today.  Thank you for looking at it.

**Comment by sadasu** [comment]
@damdo, we are not making any changes to the k8s service API. Based on upstream feedback, we are open to change the implementation. We are not changing the k8s service API here or in the upstream PR.

Once we get the upstream PR merged, we will replace this implementation with what we bring from upstream.

File: pkg/providers/v1/aws.go (hunk 4 above)
Comment (nrb) [reply pkg/providers/v1/aws.go:2419]: Yeah, these changes should go in. Will incorporate into later work.

**Comment by damdo** [comment]
Thanks @sadasu 
@nrb checked directly with upstream and confirmed that no code changes should be needed to the existing upstream PR, so we can proceed with this and adjust if needed.

**Comment by sadasu** [comment]
> Thanks @sadasu @nrb checked directly with upstream and confirmed that no code changes should be needed to the existing upstream PR, so we can proceed with this and adjust if needed.

 @damdo and @nrb, thanks for doing the due diligence. Now we know that when we do bring in the upstream changes, we won't have too many surprises given a lot of testing has already happened with this implementation.
